### PR TITLE
Fix cargo build when git branch is referenced

### PR DESCRIPTION
**Pull request by Flamefire**
(created using `eb --new-pr`)

This fixes a regression introduced by https://github.com/easybuilders/easybuild-easyblocks/pull/3483 when handling cargo crates from git repositories after an issue was identified in https://github.com/easybuilders/easybuild-easyconfigs/pull/22450.

Previously we replaced the git repository with
```
[patch."https://github.com/ritchie46/jsonpath"]
jsonpath_lib = { path = "/dev/shm/polars/0.20.2/gfbf-2023a/easybuild_vendor/jsonpath" }
```

After that PR we included the revision to ensure the correct source is taken:
```
[source."https://github.com/ritchie46/jsonpath?rev=24eaf0b4416edff38a4d1b6b17bc4b9f3f047b4b"]
git = "https://github.com/ritchie46/jsonpath"
rev = "24eaf0b4416edff38a4d1b6b17bc4b9f3f047b4b"
replace-with = "vendored-sources"
```

However a cargo crate might specify a branch instead of a revision:
`jsonpath_lib = { version = "0.3", optional = true, git = "https://github.com/ritchie46/jsonpath", branch = "improve_compiled" }`

As the resolver is very strict it needs `branch = "improve_compiled"` under `rev = ...` in this case. The `Cargo.lock` file actually has that information but we ignored it

Solution: Change the entry in the easyconfig crates list:
```diff
-('jsonpath_lib', '0.3.0', 'https://github.com/ritchie46/jsonpath', '24eaf0b4416edff38a4d1b6b17bc4b9f3f047b4b'),
+('jsonpath_lib', '0.3.0', 'https://github.com/ritchie46/jsonpath', '24eaf0b4416edff38a4d1b6b17bc4b9f3f047b4b', 'improve_compiled'),
```

This PR implements support for that:
- The extraction (easyblock run directly) now also prints a 5th tuple element which is either `None` or the branch
- If the new crate-tuple is used (branch is None or a valid string) then the new format is used including `branch = ` if the specified branch is True-ish
- If the old crate-tuple is used (backwards compatibility) the old format is used unless the new format is required:
    - for crates with git workspaces
    - when the same repository but with different revisions is used (by different crates)   
      This is required because our old format has no way for the revision and might use the wrong for one of those.   
      Using the new format here works unless the crates referenced this with a branch which should be rare enough

I also moved the offline-setup code to an own function to make it a bit easier to handle, and some refactoring of the collection code as pylint was noting e.g. that variables from the outer space are being redefined. Also the output should now be a bit more clear.

### CODE REVIEW

File: easybuild/easyblocks/generic/cargo.py [hunk 1]
```diff
@@ -295,10 +313,18 @@ def extract_step(self):
...
 
-            is_vendor_crate = src['name'] in vendor_crates
+            # Check if the source is a vendored crate and store the crate name
+            try:
+                crate_name = vendor_crates[src['name']][0]
+            except KeyError:
+                is_vendor_crate = False
+            else:
+                is_vendor_crate = True
+                src['crate_name'] = crate_name
```
Comment (lexming) [inline easybuild/easyblocks/generic/cargo.py:323]: what about:
```suggestion
            # Check if the source is a vendored crate and store the crate name
            is_vendor_crate = src['name'] in vendor_crates
            if is_vendor_crate:
                src['crate_name'] = crate_name
```

File: easybuild/easyblocks/generic/cargo.py [hunk 2]
```diff
@@ -270,11 +283,16 @@ def extract_step(self):
         for src in self.src:
             # Check for git crates, `git_key` will be set to a true-ish value for those
             try:
-                crate_name, _, git_repo, rev = vendor_crates[src['name']]
+                crate = vendor_crates[src['name']]
+                if len(crate) == 5:
+                    crate_name, _, git_repo, rev, branch = crate
+                else:
+                    crate_name, _, git_repo, rev = crate
+                    branch = self.BRANCH_MISSING
```
Comment (lexming) [inline easybuild/easyblocks/generic/cargo.py:291]: Maybe we can simplify this to if we can spare the `BRANCH_MISSING` state
```suggestion
                crate = vendor_crates[src['name']]
                crate_name, _, *git_key = crate
```

File: easybuild/easyblocks/generic/cargo.py [hunk 3]
```diff
@@ -140,6 +148,8 @@ def get_checksum(src, log):
 class Cargo(ExtensionEasyBlock):
     """Support for installing Cargo packages (Rust)"""
 
+    BRANCH_MISSING = object()  # Used when branch information was not specified
```
Comment (lexming) [inline easybuild/easyblocks/generic/cargo.py:151]: Same question as before, can we treat crates with a missing branch in the same as crates with branch=None? I would prefer this to be optional, so not mandatory to be specified.

File: easybuild/easyblocks/generic/cargo.py [hunk 4]
```diff
@@ -339,38 +365,68 @@ def extract_step(self):
...
+            if os.path.dirname(src_dir) == self.vendor_dir:
+                # Non-workspace sources are in vendor_dir
+                # We want to specify the repo using the revision and optional branch used which is most reliable.
+                if branch is not self.BRANCH_MISSING or len(repo_revs[git_repo]) > 1:
+                    self.log.debug("Writing config.toml entry for git repo: %s branch %s, rev %s",
+                                   git_repo, branch, rev)
+                    branch_info = 'branch = "%s"\n' % branch if branch else '\n'
                     write_file(config_toml,
-                               CONFIG_TOML_SOURCE_GIT.format(url=git_repo, rev=rev),
+                               CONFIG_TOML_SOURCE_GIT.format(url=git_repo, rev=rev) + branch_info,
```
Comment (lexming) [inline easybuild/easyblocks/generic/cargo.py:409]: I prefer to have a different template for the repos with branch information. This currently splits the format of that configuration in two distant parts of the code (template in line 67, branch spec in line 407), which is not clear.

```suggestion
                               CONFIG_TOML_SOURCE_GIT_BRANCH.format(url=git_repo, rev=rev, branch=branch),
```

File: easybuild/easyblocks/generic/cargo.py [hunk 5]
```diff
@@ -55,12 +57,18 @@
 
 """
 
+CONFIG_TOML_PATCH_GIT = """
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply easybuild/easyblocks/generic/cargo.py:60]: Yes this is for backward compatibility.   
There is no good way of knowing that except for checking the extracted source of each of them. IIRC a `grep branch= Cargo.toml` should do, but still a bit of work.    
The change isn't so old so even though we ran into the issue just once doesn't mean only few being affected.

If we can drop that, it would make it much nicer: We can have `(name, version, rev)` and `(name, version, rev, branch)` in the easyconfigs instead of `(name, version, rev, None)` when branch is not required.

So I'm very much open for that especially as it removes a chunk of...tricky code.

So I'm very much inclined to do that and drop backwards compatibility of some(!) easyconfigs.

Maybe we can do a `eb --stop=fetch` on "enough" cargo easyconfigs, unpack each archive and check for branch info to verify.

File: easybuild/easyblocks/generic/cargo.py (hunk 2 above)
Comment (Flamefire) [reply easybuild/easyblocks/generic/cargo.py:291]: I didn't know you can have `*git_key` here, thanks! But that's too loose. Rather:
```suggestion
                crate_name, _, repo, rev, *branch = crate
```

Because we do need the repo and revision anyway and want it to fail for `(name, version)` tuples. Only downside is the branch will be `[]` which we should change to `None` below.

File: easybuild/easyblocks/generic/cargo.py (hunk 1 above)
Comment (Flamefire) [reply easybuild/easyblocks/generic/cargo.py:323]: That is missing the `crate_name` retrieval.   
I wanted to stay true to the Python paradigm of "ask for forgiveness not permission" and avoid the double (quadruple) lookup by (not) having this in the last line: `src['crate_name'] = vendor_crates[src['name']][0]`

File: easybuild/easyblocks/generic/cargo.py (hunk 4 above)
Comment (Flamefire) [reply easybuild/easyblocks/generic/cargo.py:409]: Agreed. My thinking here was: We have a template with url and rev keys and if we have another one it would have url, rev, branch keys. So now the knowledge about the existing keys is non-local which isn't much better.

So with your suggestion we'd have `CONFIG_TOML_SOURCE_GIT` and `CONFIG_TOML_SOURCE_GIT_BRANCH` and decide on `if branch` which one to use. So something like:

```
template = CONFIG_TOML_SOURCE_GIT_BRANCH if branch else CONFIG_TOML_SOURCE_GIT
write_file(config_toml, template.format(url=git_repo, rev=rev, branch=branch))
```
`format` doesn't care about unused args, so that should be the cleanest.

File: easybuild/easyblocks/generic/cargo.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Flamefire) [reply easybuild/easyblocks/generic/cargo.py:60]: Better read the Cargo.lock because every crate referenced might reference a branch and the Cargo.lock does have that information as that is how we can get it during the collection part (when running the easyblock as a file) Otherwise you'D need to read every Cargo.toml

What I'm worried about that approach: We do the collection when creating the easyconfig and can assume Python packages such as PyToml to be available. We cannot do that when installing the easyconfig. And using RegEx or so to implement a rudimentary TOML parser makes it rather more complex. Likely a lot of edgecases there...

Then better the approach with having an optional branch in the easyconfig, dropping the old "redirect-whole-repo"-approach and sacrifice the few easyconfigs that actually require the branch  for a simpler easyblock.

[the remark this answers is not in the document]
Comment (lexming) [reply easybuild/easyblocks/generic/cargo.py:60]: > Otherwise you'D need to read every Cargo.toml

Yes :smile: 

> using RegEx or so to implement a rudimentary TOML parser

That too :smile: 

I chose to use `Cargo.toml` instead of `Cargo.lock` because it's format is much easier to parse, with a dict-like syntax.

Please check https://github.com/Flamefire/easybuild-easyblocks/pull/9

**Comment by Flamefire** [comment]
For visibility I'd like to discuss the alternatives we have so far here:

1) Have tuples of `(name, version, rev)` and `(name, version, rev, branch)` (optional branch) and use the (new) `[source."{url}?rev={rev}"]`entries with optional branch.
2) Have only `(name, version, rev)` and detect required branches by searching **all** `Cargo.toml` files if any specifies a branch. (PR by @lexming )

Advantage of 1) is that it is the flexibility: The code to print out the list can depend on e.g. the `toml` package, newer Python versions etc. as it doesn't need to run in the same constraints as the build. Easyconfig authors can change/fix issues that code introduced, e.g. manually changing/adding branch information. However it breaks some easyconfigs that exist, as seen in the polars case.   
Advantage of 2) is the easier/cleaner format in EasyConfigs and the backwards compatibility. However there might be corner-cases the parsing code did not anticipate and then fails, but that should be caught in test builds I guess. But the logic is/has to be limited to not re-implement a (full) TOML-parser and might fail or return wrong results. E.g. there are [packages](https://github.com/easybuilders/easybuild-easyconfigs/pull/21638/files#diff-2c4c5f73af66176ce0c91c8f1cb5a0b414907b3491d12c7c59f2098f29c214b0R187-R189) that use a dependency of the same name with different versions, although in that case not using git but crates.io. That might be hard to handle without a full parser.

For reference a few formats from cargo docs [1](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html) and [2](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html) (not valid, just using the same crate as example):

Cargo.toml:
```
[dependencies]
regex = { git = "https://github.com/rust-lang/regex.git" }
regex = { git = "https://github.com/rust-lang/regex.git", rev = "9f9f693" }
regex = { git = "https://github.com/rust-lang/regex.git", branch = "next" }
regex = { git = "https://github.com/rust-lang/regex.git", tag = "1.10.3" }
regex = { version = "1.10.3", git = "https://github.com/rust-lang/regex.git", branch = "next" }
some-crate = { version = "1.0", registry = "my-registry" }
time = "0.1.12"
```

Cargo.lock:
```
[[package]]
name = "regex"
version = "1.5.0"
source = "git+https://github.com/rust-lang/regex.git#9f9f693768c584971a4d53bc3c586c33ed3a6831"
[[package]]
name = "regex"
version = "1.5.0"
source = "git+https://github.com/rust-lang/regex.git&rev=9f9f693768c584971a4d53bc3c586c33ed3a6831#9f9f693768c584971a4d53bc3c586c33ed3a6831"
[[package]]
name = "regex"
version = "1.5.0"
source = "git+https://github.com/rust-lang/regex.git&branch=next#9f9f693768c584971a4d53bc3c586c33ed3a6831"
```

Given the different additional keys in `Cargo.toml` files the approach 2) might be extendable to that too.    
I'm also not sure that different versions of the same crate when using git repositories are possible but I assume so as the (transitive) dependencies might use whatever they like so I'm quite sure it can happen in practice

**Comment by lexming** [comment]
That's a very good summary.

* I'm against approach 1) mainly because the `branch` key has no role in fetching the crate sources in EasyBuild. We want a URL and a commit, always. EasyBuild will not fetch the HEAD of any branch, as that cannot be checksumed. So having the branch info in `crates` is just noise from that perspective.

* Approach 2) has some risks in parsing the toml as you point out. I agree there. That exmaple with 3 versions of `itertools` is scary. But the format in `Cargo.toml` files is well structured enough to have a reliable parsing of the `branch` key with regexes.
  We should keep in mind that this is only done for crates taken from git repos, which are quite rare on their own. Crates with git repos with branches are even more rare. So having a case with multiple of those of the same package is a very exotic case. And the code in the easyblock can easily be improved whwnever any of those complicated cases show up.

**Comment by Flamefire** [comment]
This should now be ready. I'd like to unify workspace and non-workspace git crates (see https://github.com/Flamefire/easybuild-easyblocks/pull/9#discussion_r1986760244) but my easy attempt failed as because in jsonpath/wasm it expects also a `.cargo-checksums.toml` in addition to the main folder created by cloning  https://github.com/ritchie46/jsonpath

I assume it needs one checksum file for every Cargo.toml so the whole workspace-parsing code can be replaced by a glob.

**Comment by lexming** [comment]
Test report by @lexming
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _modkit-0.3.3-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _mfqe-0.5.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _maturin-1.1.0-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _tiktoken-0.7.0-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _kyber-0.4.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _tokenizers-0.15.2-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _juliaup-1.17.9-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _jiter-0.4.1-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _Longshot-1.0.0-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _skani-0.2.2-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _smafa-0.8.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _rustworkx-0.15.1-gfbf-2023a.eb_ 
 * **SUCCESS** _Safetensors-0.4.3-gfbf-2023a.eb_ 
 * **SUCCESS** _cramino-0.14.5-GCC-12.3.0.eb_ 
 * **SUCCESS** _chopper-0.9.0-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _Clarabel.rs-0.7.1-gfbf-2023a.eb_ 
 * **SUCCESS** _fastparquet-2024.11.0-gfbf-2023a.eb_ 
 * **SUCCESS** _cryptography-41.0.1-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _phasius-0.2.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _DeltaLake-0.15.1-gfbf-2023a.eb_ 
 * **SUCCESS** _bcrypt-4.0.1-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _bamtofastq-1.4.1-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _pydantic-2.5.3-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _poetry-1.7.1-GCCcore-12.3.0.eb_ 

Build succeeded for 24 out of 24 (24 easyconfigs in total)
obelia - Linux Fedora Linux 41 (KDE Plasma), x86_64, 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz, Python 3.9.21
See https://gist.github.com/lexming/a7381efd7192a4702959d6dc23df714c for a full test report.

**Review by lexming — APPROVED** [review]
